File: examples/kotlin/examples/ClientQueryBuilder.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.QueryBuilder
import io.ekodb.client.types.Record
import kotlinx.coroutines.runBlocking

/**
 * Query builder example - Build complex queries with filters, sorting, and pagination
 */
fun main() = runBlocking {
    val baseUrl = System.getenv("API_BASE_URL") ?: "http://localhost:8080"
    val apiKey = System.getenv("API_BASE_KEY") ?: error("API_BASE_KEY environment variable not set")
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_query_example"
    
    println("=== ekoDB Kotlin Client - Query Builder Example ===\n")
    
    try {
        // Insert sample data
        println("=== Inserting Sample Data ===")
        val sampleData = listOf(
            Record.new().insert("name", "Alice").insert("age", 28).insert("city", "NYC").insert("score", 85),
            Record.new().insert("name", "Bob").insert("age", 35).insert("city", "LA").insert("score", 92),
            Record.new().insert("name", "Charlie").insert("age", 22).insert("city", "NYC").insert("score", 78),
            Record.new().insert("name", "Diana").insert("age", 31).insert("city", "Chicago").insert("score", 95),
            Record.new().insert("name", "Eve").insert("age", 27).insert("city", "LA").insert("score", 88)
        )
        
        client.batchInsert(collection, sampleData)
        println("✓ Inserted ${sampleData.size} records\n")
        
        // Example 1: Simple equality query
        println("=== Query 1: Equality (city = 'NYC') ===")
        val query1 = QueryBuilder.new()
            .eq("city", "NYC")
            .build()
        
        val results1 = client.find(collection, query1)
        println("✓ Found ${results1.size} records in NYC\n")
        
        // Example 2: Range query
        println("=== Query 2: Range (age >= 25 AND age < 32) ===")
        val query2 = QueryBuilder.new()
            .gte("age", 25)
            .lt("age", 32)
            .build()
        
        val results2 = client.find(collection, query2)
        println("✓ Found ${results2.size} records with age 25-31\n")
        
        // Example 3: Query with sorting
        println("=== Query 3: Sort by score (descending) ===")
        val query3 = QueryBuilder.new()
            .sortDesc("score")
            .limit(3)
            .build()
        
        val results3 = client.find(collection, query3)
        println("✓ Top 3 scores:")
        results3.forEach { record ->
            println("  - Score: ${record.fields["score"]}")
        }
        println()
        
        // Example 4: Complex query with multiple conditions
        println("=== Query 4: Complex (score > 80 AND age >= 25) ===")
        val query4 = QueryBuilder.new()
            .gt("score", 80)
            .gte("age", 25)
            .sortAsc("age")
            .build()
        
        val results4 = client.find(collection, query4)
        println("✓ Found ${results4.size} high-scoring adults\n")
        
        // Example 5: IN query
        println("=== Query 5: IN (city IN ['NYC', 'LA']) ===")
        val query5 = QueryBuilder.new()
            .inArray("city", listOf("NYC", "LA"))
            .build()
        
        val results5 = client.find(collection, query5)
        println("✓ Found ${results5.size} records in NYC or LA\n")
        
        // Example 6: Pagination
        println("=== Query 6: Pagination (skip 2, limit 2) ===")
        val query6 = QueryBuilder.new()
            .sortAsc("name")
            .skip(2)
            .limit(2)
            .build()
        
        val results6 = client.find(collection, query6)
        println("✓ Page 2 (2 records):")
        results6.forEach { record ->
            println("  - ${record.fields["name"]}")
        }
        println()
        
        // Example 7: Contains (substring search)
        println("=== Query 7: Contains (name contains 'a') ===")
        val query7 = QueryBuilder.new()
            .contains("name", "a")
            .build()
        
        val results7 = client.find(collection, query7)
        println("✓ Found ${results7.size} names containing 'a'\n")
        
    } finally {
        // Cleanup
        println("=== Cleanup ===")
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n=== Example Complete ===")
    }
}

File: examples/rust/examples/client_query_builder.rs
//! Query Builder Example
//!
//! This example demonstrates the QueryBuilder for constructing complex queries
//! with operators, sorting, pagination, and joins.
//!
//! Prerequisites:
//! - Run the ekoDB server: `make run`
//!
//! Run with: `cargo run --example client_query_builder`

use ekodb_client::{Client, QueryBuilder, Record};
use serde_json::json;
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Load environment variables
    dotenv::dotenv().ok();

    println!("=== ekoDB Query Builder Example ===\n");

    // Create client
    let client = Client::builder()
        .base_url(
            std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string()),
        )
        .api_key(std::env::var("API_BASE_KEY")?)
        .build()?;

    let collection = "query_builder_client_rust";

    // Step 1: Insert sample data
    println!("=== Inserting Sample Data ===");

    let users = vec![
        ("Alice", 25, "alice@example.com", "active", 1500),
        ("Bob", 30, "bob@example.com", "active", 2000),
        ("Charlie", 35, "charlie@example.com", "inactive", 1000),
        ("David", 28, "david@example.com", "active", 1800),
        ("Eve", 32, "eve@example.com", "pending", 1200),
    ];

    let user_count = users.len();
    for (name, age, email, status, score) in users {
        let mut user = Record::new();
        user.insert("name", name);
        user.insert("age", age);
        user.insert("email", email);
        user.insert("status", status);
        user.insert("score", score);
        client.insert(collection, user, None).await?;
    }
    println!("✓ Inserted {} users\n", user_count);

    // Step 2: Simple equality query
    println!("=== Simple Equality Query ===");
    let query = QueryBuilder::new().eq("status", "active").build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} active users", results.len());
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 3: Range query
    println!("=== Range Query (age >= 28 AND age < 35) ===");
    let query = QueryBuilder::new().gte("age", 28).lt("age", 35).build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users in age range", results.len());
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 4: IN operator
    println!("=== IN Operator ===");
    let query = QueryBuilder::new()
        .in_array("status", vec![json!("active"), json!("pending")])
        .build();

    let results = client.find(collection, query, None).await?;
    println!(
        "✓ Found {} users with status active or pending",
        results.len()
    );
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 5: NOT IN operator
    println!("=== NOT IN Operator ===");
    let query = QueryBuilder::new()
        .nin("status", vec![json!("inactive")])
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users not inactive", results.len());
    println!();

    // Step 6: String pattern matching
    println!("=== String Pattern Matching ===");
    let query = QueryBuilder::new()
        .contains("email", "@example.com")
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users with @example.com email", results.len());
    println!();

    // Step 7: Regex query
    println!("=== Regex Query ===");
    let query = QueryBuilder::new()
        .regex("name", "^[A-C]") // Names starting with A, B, or C
        .build();

    let results = client.find(collection, query, None).await?;
    println!(
        "✓ Found {} users with names starting with A-C",
        results.len()
    );
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 8: Complex query with multiple conditions
    println!("=== Complex Query (active AND age >= 28 AND score > 1500) ===");
    let query = QueryBuilder::new()
        .eq("status", "active")
        .gte("age", 28)
        .gt("score", 1500)
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users matching all conditions", results.len());
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 9: OR query
    println!("=== OR Query ===");
    let query = QueryBuilder::new()
        .or(vec![
            json!({
                "type": "Condition",
                "content": {
                    "field": "age",
                    "operator": "Lt",
                    "value": 28
                }
            }),
            json!({
                "type": "Condition",
                "content": {
                    "field": "age",
                    "operator": "Gt",
                    "value": 32
                }
            }),
        ])
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users with age < 28 OR age > 32", results.len());
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 10: Sorting
    println!("=== Sorted Query (by score descending) ===");
    let query = QueryBuilder::new().sort_desc("score").limit(3).build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Top 3 users by score:");
    for (i, user) in results.iter().enumerate() {
        println!("  {}. {:?}", i + 1, user.get("name"));
    }
    println!();

    // Step 11: Pagination
    println!("=== Pagination (page 2, size 2) ===");
    let query = QueryBuilder::new()
        .sort_asc("name")
        .page(1, 2) // Page 1 (0-indexed), 2 items per page
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Page 2 results:");
    for user in &results {
        println!("  - {:?}", user.get("name"));
    }
    println!();

    // Step 12: NOT operator
    println!("=== NOT Operator ===");
    let query = QueryBuilder::new()
        .not(json!({
            "type": "Condition",
            "content": {
                "field": "status",
                "operator": "Eq",
                "value": "inactive"
            }
        }))
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users NOT inactive", results.len());
    println!();

    // Step 13: Complex chained query
    println!("=== Complex Chained Query ===");
    let query = QueryBuilder::new()
        .eq("status", "active")
        .gte("age", 25)
        .contains("email", "@example.com")
        .sort_desc("score")
        .limit(5)
        .bypass_cache(true)
        .build();

    let results = client.find(collection, query, None).await?;
    println!("✓ Found {} users with all conditions", results.len());
    println!();

    // Cleanup: Delete the collection
    println!("=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection\n");

    println!("✓ All query builder operations completed successfully");

    Ok(())
}
